### PR TITLE
Support for kick/ban commands on Gitter

### PR DESCRIPTION
**Pull request by Half-Shot**
image

This PR allows Gitter admins to kick or ban users from their rooms.

### CODE REVIEW

**Comment by Half-Shot** [comment]
Oops, forgot to auth the command :fearful:

File: lib/BridgedRoom.js [hunk 1]
```diff
@@ -427,6 +432,74 @@ BridgedRoom.prototype.onGitterMessage = function(message) {
...
+    const command = commandParts[0].toLowerCase();
+    const commandArgs = commandParts.slice(1);
+
+    const roomIds = new Array(this._linkedMatrixRoomIds);
+    if (this._portalMatrixRoomId) {
+        roomIds.push(this._portalMatrixRoomId);
+    }
+
+    let response = "Command not understood.";
+    if (command === "kick" || command === "ban" && commandArgs.length === 0) {
```
Comment (anoadragon453) [inline lib/BridgedRoom.js]:
```suggestion
    if ((command === "kick" || command === "ban") && commandArgs.length === 0) {
```

File: lib/BridgedRoom.js [hunk 2]
```diff
@@ -427,6 +432,74 @@ BridgedRoom.prototype.onGitterMessage = function(message) {
...
+        response = `\`${command}\` needs a name parameter. Specify a matrix user ID (@foo:bar) or displayname`;
+    } else if (command === "kick" || command === "ban") {
+        const givenUser = commandArgs[0];
+        let userId;
+        const isUserId = IS_USERID.match(userId);
```
Comment (anoadragon453) [inline lib/BridgedRoom.js]: Why would this ever return `true` if you initialize it the line before?

File: lib/BridgedRoom.js [hunk 3]
```diff
@@ -427,6 +432,74 @@ BridgedRoom.prototype.onGitterMessage = function(message) {
...
+            log.debug(`Matching '${givenUser}' as user id`);
+        }
+        // Because we can't ask the gitter admin to target a single room, we run this across
+        // all rooms. If we manage to do it once, consider it a success.
+        let actioned = false;
+        for (const roomId of roomIds) {
+            let joinedMembers;
+            try {
+                joinedMembers = await this._main.getJoinedMembers(roomId);
+            } catch (ex) {
```
Comment (anoadragon453) [inline lib/BridgedRoom.js]: Should we log this exception?

File: lib/BridgedRoom.js [hunk 4]
```diff
@@ -427,6 +432,74 @@ BridgedRoom.prototype.onGitterMessage = function(message) {
...
+                log.warn(`Request to joined_members failed in ${roomId}. Ignoring room`);
+                continue;
+            }
+
+            // This must be run for each room.
+            if (!isUserId) {
+                // joined_members surprisngly returns per room profiles, then user profiles in that order
```
Comment (anoadragon453) [inline lib/BridgedRoom.js]:
```suggestion
                // joined_members surprisingly returns per room profiles, then user profiles in that order
```

File: lib/BridgedRoom.js [hunk 5]
```diff
@@ -427,6 +432,74 @@ BridgedRoom.prototype.onGitterMessage = function(message) {
...
+            } else if (joinedMembers[userId]) {
+                log.debug(`${userId} was not found in ${roomId}`);
+            }
+
+            try {
+                await this._bridge.getBotIntent()[command](roomId, userId, `Requested by ${fromUser.username}`);
+                log.info(`Actioned ${command} against ${userId} in ${roomId}`);
+                actioned = true;
+            } catch (ex) {
+                // Report these errors to Gitter?
```
Comment (anoadragon453) [inline lib/BridgedRoom.js]:
```suggestion
                // TODO: Report these errors to Gitter?
```

?

File: lib/BridgedRoom.js [hunk 6]
```diff
@@ -427,6 +432,74 @@ BridgedRoom.prototype.onGitterMessage = function(message) {
...
+                log.debug(`Failed to ${command} ${userId} in ${roomId}:`, ex);
+            }
+        }
+
+        response = actioned ? "Successfully completed action against user" : "User could not be removed";
```
Comment (anoadragon453) [inline lib/BridgedRoom.js]:
```suggestion
        response = actioned ? `Successfully completed ${command} against user` : "User could not be removed";
```

File: lib/BridgedRoom.js [hunk 7]
```diff
@@ -434,6 +446,85 @@ BridgedRoom.prototype.onGitterMessage = function(message) {
...
+    const IS_USERID = /^@.+:.+$/; // Lazy user id regex.
+    if (message === "!matrix") {
+        message = "!matrix help";
+    }
+
+    // Determine if the user is an admin.
+    await this._listGitterUsers();
+
+    if (!this._gitterAdminList.has(fromUser.username)) {
+        await this._main.getRelayGhost().send(this._gitterRoomName, "Only admin's may use the !matrix command");
```
Comment (anoadragon453) [inline lib/BridgedRoom.js]:
```suggestion
        await this._main.getRelayGhost().send(this._gitterRoomName, "Only admins may use the !matrix command");
```

File: lib/BridgedRoom.js [hunk 8]
```diff
@@ -434,6 +446,85 @@ BridgedRoom.prototype.onGitterMessage = function(message) {
...
+        return;
+    }
+
+
+    const commandParts = message.substr("!matrix".length).split(" ");
```
Comment (anoadragon453) [inline lib/BridgedRoom.js]: image

I don't think this does what you want it to do.

```suggestion
    const commandParts = message.split(" ");
```

Alternatively:

```
let commandArgs = message.split(" ");
const command = commandArgs.shift().toLowerCase();
```

File: lib/BridgedRoom.js [hunk 9]
```diff
@@ -434,6 +446,85 @@ BridgedRoom.prototype.onGitterMessage = function(message) {
...
+    const command = commandParts[0].toLowerCase();
+    const commandArgs = commandParts.slice(1);
+
+    const roomIds = new Array(this._linkedMatrixRoomIds);
+    if (this._portalMatrixRoomId) {
+        roomIds.push(this._portalMatrixRoomId);
+    }
+
+    let response = "Command not understood.";
```
Comment (anoadragon453) [inline lib/BridgedRoom.js]: Some of your responses have punctuation and others don't.

File: lib/BridgedRoom.js [hunk 10]
```diff
@@ -434,6 +446,85 @@ BridgedRoom.prototype.onGitterMessage = function(message) {
...
+        // Because we can't ask the gitter admin to target a single room, we run this across
+        // all rooms. If we manage to do it once, consider it a success.
+        let actioned = false;
+        for (const roomId of roomIds) {
+            let joinedMembers;
+            try {
+                joinedMembers = await this._main.getJoinedMembers(roomId);
+            } catch (ex) {
+                log.warn(`Request to joined_members failed in ${roomId}. Ignoring room.`);
+                log.debug("Exception:",ex);
```
Comment (anoadragon453) [inline lib/BridgedRoom.js]:
```suggestion
                log.debug("Exception:", ex);
```

File: lib/BridgedRoom.js (hunk 9 above)
Comment (Half-Shot) [reply lib/BridgedRoom.js]: I think it's consistent now

File: lib/BridgedRoom.js [hunk 11]
```diff
@@ -434,6 +453,98 @@ BridgedRoom.prototype.onGitterMessage = function(message) {
...
+                log.info(`Actioned ${command} against ${userId} in ${roomId}`);
+                actioned = true;
+            } catch (ex) {
+                // TODO: Report these errors to Gitter?
+                log.debug(`Failed to ${command} ${userId} in ${roomId}:`, ex);
+            }
+        }
+
+        if (actioned) {
+            response = null;
```
Comment (anoadragon453) [inline lib/BridgedRoom.js]: Shouldn't we just return here? We can then also get rid of the below `if` statement.

File: lib/BridgedRoom.js [hunk 12]
```diff
@@ -587,10 +704,18 @@ BridgedRoom.prototype._listGitterUsers = function() {
             batch.forEach((u) => {
                 if (u.id === myId) return;
                 names.add(u.username.toLowerCase());
+                console.log(u);
```
Comment (anoadragon453) [inline lib/BridgedRoom.js]: Debugging

**Pull request merged**